### PR TITLE
PR: Migrate the Outline Explorer to use LSP information

### PR DESCRIPTION
**Pull request by andfoy**
## Description of Changes
This PR migrates the Outline Explorer to use the information provided by the LSP via `textDocument/documentSymbols` call.

ezgif com-optimize

### Loading spinner

Peek 04-09-2020 18-26

This PR depends on https://github.com/palantir/python-language-server/pull/846

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @andfoy

### CODE REVIEW

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py [hunk 1]
```diff
@@ -230,7 +258,158 @@ def test_toggle_on_show_all_files(editorstack, outlineexplorer, test_files):
     editorstack.tabs.tabBar().moveTab(0, 1)
     treewidget.toggle_show_all_files(True)
     results = [item.text(0) for item in treewidget.get_visible_items()]
-    assert results == ['text1.txt', 'foo1.py', 'foo', 'foo2.py']
+    assert results == ['text1.txt', 'foo1.py', 'foo2.py']
+
+
+@pytest.mark.slow
+@pytest.mark.second
+def test_editor_outlineexplorer(qtbot, lsp_codeeditor_outline):
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_editor_and_outline.py:266]: Please add a docstring to describe what this test does.

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py [hunk 2]
```diff
@@ -230,7 +258,158 @@ def test_toggle_on_show_all_files(editorstack, outlineexplorer, test_files):
...
+        code_editor.document_did_change()
+
+    with qtbot.waitSignal(treewidget.sig_tree_updated, timeout=30000):
+        code_editor.request_symbols()
+
+    tree = trees[2]
+    root_tree = get_tree_elements(treewidget)
+    assert root_tree == tree
+
+    # Move method1
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_editor_and_outline.py:356]: I think `Move` should be `Remove` here?

File: spyder/plugins/outlineexplorer/editor.py [hunk 3]
```diff
@@ -6,6 +6,13 @@
 
 """Outline explorer editor server"""
 
+# Standard library imports
+import uuid
```
Comment (ccordoba12) [inline spyder/plugins/outlineexplorer/editor.py:10]: It seems this import is not needed here. If that's the case, please remove it.

File: spyder/plugins/outlineexplorer/tests/test_widgets.py [hunk 4]
```diff
@@ -17,50 +19,22 @@
...
-        "            pass")
+AVAILABLE_CASES = ['text']
+CASES = {
+    case: {
+        'file': osp.join(ASSETS, '{0}_{1}.py'.format(case, SUFFIX)),
+        'data': osp.join(ASSETS, '{0}_{1}.json'.format(case, SUFFIX)),
+        'tree': osp.join(ASSETS, '{0}_exp_{1}.json'.format(case, SUFFIX))
+    }
+    for case in AVAILABLE_CASES
+}
```
Comment (ccordoba12) [inline spyder/plugins/outlineexplorer/tests/test_widgets.py:37]: Why is this necessary if there's only one available case (i.e. `text`)? What other kind of cases there could be?

File: spyder/plugins/outlineexplorer/editor.py [hunk 5]
```diff
@@ -14,12 +21,13 @@ def __init__(self, editor, fname):
         super(OutlineExplorerProxyEditor, self).__init__()
         self._editor = editor
         self.fname = fname
+        self.symbol_info = {}
+        self.current_tree = IntervalTree()
```
Comment (ccordoba12) [inline spyder/plugins/outlineexplorer/editor.py:25]: Are these two attributes really necessary? They don't seem to be used elsewhere in this PR.

File: spyder/plugins/outlineexplorer/widgets.py [hunk 6]
```diff
@@ -359,24 +448,28 @@ def set_current_editor(self, editor, update):
...
             if not self.freeze:
                 self.scrollToItem(item)
                 self.root_item_selected(item)
                 self.__hide_or_show_root_items(item)
             if update:
                 self.save_expanded_state()
-                tree_cache = self.editor_tree_cache[editor_id]
-                self.populate_branch(editor, item, tree_cache)
+                # tree_cache = self.editor_tree_cache[editor_id]
+                # self.populate_branch(editor, item, tree_cache)
```
Comment (ccordoba12) [inline spyder/plugins/outlineexplorer/widgets.py:459]: If this code is not necessary anymore, please remove it.

File: spyder/plugins/outlineexplorer/widgets.py [hunk 7]
```diff
@@ -359,24 +448,28 @@ def set_current_editor(self, editor, update):
...
         else:
-            root_item = FileRootItem(editor.fname, self, editor.is_python())
+            this_root = SymbolStatus(editor.fname, None, None, editor.fname)
+            root_item = FileRootItem(editor.fname, this_root,
+                                     self, editor.is_python())
             root_item.set_text(fullpath=self.show_fullpath)
-            tree_cache = self.populate_branch(editor, root_item)
+            editor_tree = IntervalTree()
+            this_root.node = root_item
+            # tree_cache = self.populate_branch(editor, root_item)
```
Comment (ccordoba12) [inline spyder/plugins/outlineexplorer/widgets.py:468]: Remove if it's not necessary.

File: spyder/plugins/outlineexplorer/widgets.py [hunk 8]
```diff
@@ -409,29 +502,124 @@ def update_all(self):
...
+        update = self.update_tree(items, editor_id)
+        if getattr(plugin_base, "_isvisible", True) and update:
             self.save_expanded_state()
-            editor = self.current_editor
-            editor_id = editor.get_id()
             self.__do_update(editor, editor_id)
             self.restore_expanded_state()
             self.do_follow_cursor()
 
+    def merge_interval(self, parent, node):
```
Comment (ccordoba12) [inline spyder/plugins/outlineexplorer/widgets.py:521]: Please add a docstring to understand what this method does.

File: spyder/plugins/outlineexplorer/widgets.py [hunk 9]
```diff
@@ -409,29 +502,124 @@ def update_all(self):
...
     def __do_update(self, editor, editor_id):
         """
         Recalculate the and update the tree items in the Outliner for a
         given editor
         """
-        item = self.editor_items[editor_id]
+        item = self.editor_items[editor_id].node
         tree_cache = self.editor_tree_cache[editor_id]
-        self.populate_branch(editor, item, tree_cache)
+        # self.populate_branch(editor, item, tree_cache)
```
Comment (ccordoba12) [inline spyder/plugins/outlineexplorer/widgets.py:622]: Please remove if it's not necessary.

File: spyder/utils/icon_manager.py [hunk 10]
```diff
@@ -288,22 +288,39 @@
...
+    'array':                   [('mdi.alpha-a-box',), {'color': '#df2935', 'scale_factor': BIG_ATTR_FACTOR}],
+    'object':                  [('mdi.alpha-o-box',), {'color': 'yellow', 'scale_factor': BIG_ATTR_FACTOR}],
+    'key':                     [('mdi.alpha-k-box',), {'color': 'yellow', 'scale_factor': BIG_ATTR_FACTOR}],
+    'class':                   [('mdi.alpha-c-box',), {'color':'#3775a9', 'scale_factor': BIG_ATTR_FACTOR}],
+    'null':                    [('mdi.alpha-n-box',), {'color': 'magenta', 'scale_factor': BIG_ATTR_FACTOR}],
+    'enum_member':             [('mdi.alpha-e-box',), {'color': '#e69c9c', 'scale_factor': BIG_ATTR_FACTOR}],
+    'struct':                  [('mdi.alpha-s-box',), {'color':'#3775a9', 'scale_factor': BIG_ATTR_FACTOR}],
+    'event':                   [('mdi.alpha-e-box',), {'color':'orange', 'scale_factor': BIG_ATTR_FACTOR}],
+    'operator':                [('mdi.alpha-o-box',), {'color':'orange', 'scale_factor': BIG_ATTR_FACTOR}],
+    'type_parameter':          [('mdi.alpha-t-box',), {'color':'orange', 'scale_factor': BIG_ATTR_FACTOR}],
```
Comment (ccordoba12) [inline spyder/utils/icon_manager.py:323]: How did you pick up the colors of the new icons you added in this file?

File: spyder/utils/syntaxhighlighters.py [hunk 11]
```diff
@@ -504,8 +504,8 @@ def __init__(self, parent, font=None, color_scheme='Spyder'):
         # Avoid updating the outline explorer with every single letter typed
         self.outline_explorer_data_update_timer = QTimer()
         self.outline_explorer_data_update_timer.setSingleShot(True)
-        self.outline_explorer_data_update_timer.timeout.connect(
-            self.sig_outline_explorer_data_changed)
+        # self.outline_explorer_data_update_timer.timeout.connect(
+        #     self.sig_outline_explorer_data_changed)
```
Comment (ccordoba12) [inline spyder/utils/syntaxhighlighters.py:508]: Please remove these comments if they are not necessary.

**Comment by ccordoba12** [comment]
My comments from manually testing this:

* This PR is showing the imported modules in a file, which takes a lot of space (one per line) and I don't think it's really necessary. Perhaps we should show remove them? How does this work in VSCode? 

    Selección_016

* Multiline imports are shown as nested now.

    Selección_017

    Selección_018

* This PR is only showing the current file on the tree. This is a loss of a functionality because right now we are showing all open files on the Outline:

    Selección_019

* It takes several seconds (three or more) to display the contents of the tree the first time a file is opened. This is something we really need to fix before merging it.

**Comment by andfoy** [comment]
> This PR is only showing the current file on the tree. That's a loss of a functionality because right now we are showing all open files on the Outline:

If you have the option `Show all files`, then all the files will appear:

imagen

**Comment by andfoy** [comment]
> How does this work in VSCode?

It seems that variables are displayed, but imports not.

**Comment by andfoy** [comment]
> Methods and attributes of classes are displayed as variables and functions:

This is a limitation of the LSP spec, which does not consider attributes nor methods

File: spyder/plugins/editor/widgets/tests/test_editor_and_outline.py (hunk 2 above)
Comment (andfoy) [reply spyder/plugins/editor/widgets/tests/test_editor_and_outline.py:356]: It is move because of the cut operation performed below

File: spyder/plugins/outlineexplorer/tests/test_widgets.py (hunk 4 above)
Comment (andfoy) [reply spyder/plugins/outlineexplorer/tests/test_widgets.py:37]: This allows to add more test cases on the future if necessary

**Comment by ccordoba12** [comment]
Thanks @andfoy for your latest updates to this PR! The spinner interaction is working great, but we're still showing a lot of unnecessary imports:

Selección_092

I think this is something that needs to be fixed because it constitutes an important regression with respect to what we have now:

Selección_093

**Review by ccordoba12 — APPROVED** [review]
Thanks a lot @andfoy! This is a fantastic improvement!!

**Comment by CAM-Gerlach** [comment]
For the record, as requested by @ccordoba12 , some quick test results moving lines that result in Outline Explorer updates, all with default Spyder settings, on mainwindow.py starting from around line 1000 and moving for 100 lines:

Previous result without this PR on 4.x, Outline Explorer hidden: 13 lines/s
Current result with this PR, Outline Explorer hidden: 13 lines/s
Current result with this PR, Outline Explorer shown: 10 lines/s

The Outline Explorer appears to update approximately 1-2x second with perhaps 100-200 ms lag, which would explain most of the slowdown (assuming it was blocking or stealing cycles from GUI updates during most of that time).
